man/mkss.1
'\" t
.\"     Title: 
    mkss
  
.\"    Author: 
.\" Generator: DocBook XSL Stylesheets v1.76.1 <http://docbook.sf.net/>
.\"      Date: 12/19/2013
.\"    Manual: 
.\"    Source: 
.\"  Language: English
.\"
.TH "MKSS" "1" "12/19/2013" ""
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"

    mkss
   \- calculate average spectrum for spectral subtraction
.SH "SYNOPSIS"
.HP \w'\fBmkss\fR\ 'u
\fBmkss\fR [\fIoptions\fR...] {filename}
.SH "DESCRIPTION"
.PP

\fBmkss\fR
is a tool to estimate noise spectrum for spectral subtraction on Julius\&. It reads a few seconds of sound data from microphone input, calculate the average spectrum and save it to a file\&. The output file can be used as a noise spectrum data in Julius (option "\fB\-ssload\fR")\&.
.PP
The recording will start immediately after startup\&. Sampling format is 16bit, monoral\&. If outpue file already exist, it will be overridden\&.
.SH "OPTIONS"
.PP
\fB \-freq \fR \fIHz\fR
.RS 4
Sampling frequency in Hz (default: 16,000)
.RE
.PP
\fB \-len \fR \fImsec\fR
.RS 4
capture length in milliseconds (default: 3000)
.RE
.PP
\fB \-fsize \fR \fIsample_num\fR
.RS 4
frame size in number of samples (default: 400)
.RE
.PP
\fB \-fshift \fR \fIsample_num\fR
.RS 4
frame shift in number of samples (default: 160)
.RE
.SH "SEE ALSO"
.PP

\fB julius \fR( 1 )
.SH "COPYRIGHT"
.PP
Copyright (c) 1991\-2013 Kawahara Lab\&., Kyoto University
.PP
Copyright (c) 1997\-2000 Information\-technology Promotion Agency, Japan
.PP
Copyright (c) 2000\-2005 Shikano Lab\&., Nara Institute of Science and Technology
.PP
Copyright (c) 2005\-2013 Julius project team, Nagoya Institute of Technology
.SH "LICENSE"
.PP
The same as Julius\&.


ChangeLog
2008-05-17 13:06  sumomo

	* libjulius/src/: instance.c, m_fusion.c, m_options.c: Now GMM
	  module uses the analysis parameter of the given AM when "-AM_GMM"
	  is not explicitly specified.	Now output error when GMM has
	  different sampling rate from AM.

2008-05-13 15:32  sumomo

	* jcontrol/jcontrol.c: Flush output after each message.

2008-05-13 15:32  sumomo

	* libsent/src/util/readfile.c: Fixed file name prompt to ignore
	  empty input correctly.

2008-05-09 15:00  sumomo

	* libjulius/src/m_usage.c: Grouped the usage messages by the
	  belonging modules.

2008-05-09 14:59  sumomo

	* libjulius/src/m_options.c: New option "-usepower" to use power
	  instead of magnitude in filterbank analysis.

2008-05-09 14:58  sumomo

	* libsent/: include/sent/mfcc.h, src/hmminfo/read_binhmm.c,
	  src/hmminfo/write_binhmm.c, src/wav2mfcc/mfcc-core.c: Added
	  support for USEPOWER in filterbank analysis.

2008-05-09 14:57  sumomo

	* libsent/src/wav2mfcc/para.c: Add support for USEPOWER.  Fixed
	  -htkconf to assume SOURCERATE=625 (16kHz) if not exist.

2008-05-08 14:45  sumomo

	* libjulius/: include/julius/extern.h, src/adin-cut.c,
	  src/m_adin.c: Add tiny value check for frequency.

2008-05-08 14:44  sumomo

	* libsent/src/adin/adin_mic_linux_alsa.c: Fixed failure to startup
	  when using multiple sound cards on ALSA.

2008-05-08 14:43  sumomo

	* libsent/src/adin/adin_file.c: Fixed segfault when failed to open
	  audio file.

2008-04-27 22:06  sumomo

	* libjulius/: include/julius/jconf.h, src/default.c, src/m_info.c,
	  src/m_options.c, src/m_usage.c, src/search_bestfirst_main.c: Add
	  an option "-fallback1pass" to use the 1st pass result as final
	  result when the 2nd pass fails.  This was the default on 3.x.

2008-04-27 22:06  sumomo

	* libjulius/src/version.c.in: Update copyright output.

2008-03-28 13:52  sumomo

	* libsent/src/adin/adin_mic_linux_alsa.c: Fix ALSA A/D-in to use
	  device other than "default".

2008-03-28 12:57  sumomo

	* julius/record.c: Fixed record sometimes fails.

2008-03-28 12:51  sumomo

	* libjulius/src/adin-cut.c: Moved CALLBACK_ADIN_TRIGGERED to adin
	  thread.

2008-03-25 17:12  sumomo

	* libjulius/src/pass1.c: Fixed GMM rejection not working well on
	  microphone input (enbuged on 4.0.1).

2008-03-21 11:59  sumomo

	* libsent/src/wav2mfcc/para.c: Add warning when HTK Config given by
	  "-htkconf" has unsupported values.

2008-03-20 18:08  sumomo

	* libsent/: include/sent/ngram2.h, include/sent/util.h,
	  src/ngram/ngram_access.c, src/ngram/ngram_compact_context.c,
	  src/ngram/ngram_read_arpa.c, src/ngram/ngram_read_bin.c,
	  src/ngram/ngram_util.c, src/ngram/ngram_write_bin.c,
	  src/util/endian.c: Update N-gram codes to (hopefully) allow 64bit
	  index in the future.

2008-03-20 18:06  sumomo

	* julius/: charconv_win32.c, libjcode/jlib.h, libjcode/libjcode.c:
	  Fixed euc-jp conversion not working on some string.

2008-03-18 02:00  sumomo

	* libjulius/src/: adin-cut.c, recogmain.c: Fixed
	  CALLBACK_EVENT_SPEECH_* call timing when -spsegment/VAD enabled.

2008-03-18 01:59  sumomo

	* libjulius/src/version.c.in, libsent/src/util/confout.c: Fix
	  typos.

2008-03-18 01:58  sumomo

	* julius/main.c, julius/output_stdout.c,
	  libjulius/include/julius/global.h, libjulius/src/callback.c,
	  libjulius/src/m_options.c, libjulius/src/m_usage.c:
	  "-callbackdebug" moved from Julius to JuliusLib.

2008-03-14 22:28  sumomo

	* libsent/src/ngram/ngram_compact_context.c: Fixed a bug that
	  prevents reading big N-gram.

2008-03-12 15:43  tag release_4_0_1

2008-03-12 15:43  sumomo

	* 00readme-ja.txt, 00readme.txt, ChangeLog, LICENSE.txt,
	  Release-ja.txt, Release.txt, libjulius/configure,
	  libjulius/configure.in, libsent/configure, libsent/configure.in,
	  support/build-all.sh: Release of 4.0.1.

2008-03-12 14:48  sumomo

	* libjulius/src/pass1.c: Fixed compilation failure enbuged at last
	  fix...

2008-03-10 11:39  sumomo

	* libjulius/src/: pass1.c, realtime-1stpass.c, recogmain.c: Fix CMN
	  not working well on spsegment/GMM-VAD/decoder-VAD mode.

2008-03-02 02:10  sumomo

	* libsent/: configure, configure.in, include/sent/config.h.in,
	  src/adin/adin_mic_linux_alsa.c: Fixed ALSA recording to work.
	  Many codes are updated.

2008-02-21 13:10  sumomo

	* libjulius/src/beam.c: Word recognition results are not correctly
	  sorted when using "-output N".

2008-02-18 23:35  sumomo

	* julius/module.c: Added a module command "GRAMINFO" to return
	  grammar info.

2008-02-18 23:32  sumomo

	* libjulius/src/m_fusion.c: Now -spsegment detects error when pause
	  word is not defined in dictionary.

2008-02-15 03:26  sumomo

	* libsent/src/phmm/mkwhmm.c: Fixed bug of immediate stop when no sp
	  model in multi-path mode.

2008-02-15 03:25  sumomo

	* libjulius/src/: m_info.c, search_bestfirst_v1.c,
	  search_bestfirst_v2.c, word_align.c: Fixed hungup when sp model
	  not exist on AM in multipath.

2008-02-14 13:44  sumomo

	* libjulius/: include/julius/extern.h, src/multi-gram.c: Changes
	  for better grammar handling: - multigram_add() returns the newly
	  assigned grammar id.	- a function is added to get grammar id
	  from name string.  - modified STAT message to output grammar name
	  together with id.

2008-02-14 13:41  sumomo

	* julius/module.c: Extended ACTIVATEGRAM, DEACTIVATEGRAM, DELGRAM
	  to specify grammar by name.

2008-02-13 17:49  sumomo

	* libjulius/include/julius/extern.h: Updated for some function
	  changes.

2008-02-13 17:48  sumomo

	* libjulius/src/: outprob_style.c, search_bestfirst_v1.c,
	  search_bestfirst_v2.c: Global variables and work area has been
	  moved to RecogProcess and StackDecode.

2008-02-13 17:47  sumomo

	* libjulius/src/wchmm.c: qsort function switched to qsort_reentrant
	  in libsent for modularity.

2008-02-13 17:46  sumomo

	* libjulius/src/confnet.c: Updated for being re-entrant.

2008-02-13 17:45  sumomo

	* libjulius/src/: m_jconf.c, m_options.c, ngram_decode.c,

mkss/mkss.c
/*
 * Copyright (c) 2002-2013 Kawahara Lab., Kyoto University
 * Copyright (c) 2002-2005 Shikano Lab., Nara Institute of Science and Technology
 * Copyright (c) 2005-2013 Julius project team, Nagoya Institute of Technology
 * All rights reserved
 */

/*
 * mkss --- compute average spectrum of mic input for SS in Julius
 *
 * $Id: mkss.c,v 1.9 2013/12/19 16:26:22 sumomo Exp $
 *
 */

#include <julius/juliuslib.h>
#include <sys/stat.h>

static int fd = -1;		/* file descriptor for output */

static char *filename = NULL;	/* output file name */
static boolean stout = FALSE;	/* true if output to stdout ("-") */
static int sfreq;	/* sampling frequency */
static int slen  = 3000;	/* record length in msec */

/* parameter for SS */
static SP16 *speech;
static int speechnum;
static int samples;

static boolean
opt_help(Jconf *jconf, char *arg[], int argnum)
{
  fprintf(stderr, "mkss --- compute averate spectrum of mic input for SS\n");
  fprintf(stderr, "Usage: mkss [options..] filename\n");
  fprintf(stderr, "    [-freq frequency]    sampling freq in Hz   (%d)\n", jconf->am_root->analysis.para_default.smp_freq);
  fprintf(stderr, "    [-len msec]          record length in msec (%d)\n", slen);
  fprintf(stderr, "    [-fsize samplenum]   window size           (%d)\n", jconf->am_root->analysis.para_default.framesize);
  fprintf(stderr, "    [-fshift samplenum]  frame shift           (%d)\n", jconf->am_root->analysis.para_default.frameshift);
  fprintf(stderr, "    [-zmean]             enable zmean         (off)\n");
  fprintf(stderr, "    [-zmeanframe]        frame-wise zmean     (off)\n");
  fprintf(stderr, "Library configuration: ");
  confout_version(stderr);
  confout_audio(stderr);
  confout_process(stderr);
  fprintf(stderr, "\n");
  exit(1);			/* exit here */
  return TRUE;
}

static boolean
opt_freq(Jconf *jconf, char *arg[], int argnum)
{
  jconf->amnow->analysis.para.smp_freq = atoi(arg[0]);
  jconf->amnow->analysis.para.smp_period = freq2period(jconf->amnow->analysis.para.smp_freq);
  return TRUE;
}
static boolean
opt_len(Jconf *jconf, char *arg[], int argnum)
{
  slen = atoi(arg[0]);
  return TRUE;
}

static int
adin_callback(SP16 *now, int len, Recog *recog)
{
  int num;
  int ret;

  /* store recorded data up to samples */
  if (speechnum + len > samples) {
    num = samples - speechnum;
    /* stop recording */
    ret = 1;
  } else {
    num = len;
    /* continue recording */
    ret = 0;
  }
  memcpy(&(speech[speechnum]), now, num * sizeof(SP16));

  if (speechnum / sfreq != (speechnum + num) / sfreq) {
    fprintf(stderr, "|");
  } else {
    fprintf(stderr, ".");
  }

  speechnum += num;

  return(ret);
}

static int x;
static int sslen;

int
main(int argc, char *argv[])
{
  Recog *recog;
  Jconf *jconf;
  float *ss;
  MFCCWork *wrk;

  /* create instance */
  recog = j_recog_new();
  jconf = j_jconf_new();
  recog->jconf = jconf;

  /* set application-specific additional options */
  j_add_option("-freq", 1, 1, "sampling freq in Hz", opt_freq);
  j_add_option("-len", 1, 1, "record length in msec", opt_len);
  j_add_option("-h", 0, 0, "display this help", opt_help);
  j_add_option("-help", 0, 0, "display this help", opt_help);
  j_add_option("--help", 0, 0, "display this help", opt_help);

  /* when no argument, output help and exit */
  if (argc <= 1) {
    opt_help(jconf, NULL, 0);
    return 0;
  }

  /* regard last arg as filename */
  if (strmatch(argv[argc-1], "-")) {
    stout = TRUE;
  } else {
    filename = argv[argc-1];
  }

  /* set default as same as "-input mic" */
  jconf->input.type = INPUT_WAVEFORM;
  jconf->input.speech_input = SP_MIC;
  jconf->input.device = SP_INPUT_DEFAULT;
  /* process config and load them */
  if (j_config_load_args(jconf, argc-1, argv) == -1) {
    fprintf(stderr, "Error reading arguments\n");
    return -1;
  }
  /* force some default values */
  jconf->detect.silence_cut  = 0; /* disable silence cut */
  jconf->preprocess.strip_zero_sample = TRUE; /* strip zero samples */
  jconf->detect.level_thres = 0;	/* no VAD, record all */
  /* set Julius default parameters for unspecified acoustic parameters */
  apply_para(&(jconf->am_root->analysis.para), &(jconf->am_root->analysis.para_default));
  /* set some values */
  jconf->input.sfreq = jconf->am_root->analysis.para.smp_freq;
  jconf->input.period = jconf->am_root->analysis.para.smp_period;
  jconf->input.frameshift = jconf->am_root->analysis.para.frameshift;
  jconf->input.framesize = jconf->am_root->analysis.para.framesize;

  sfreq = jconf->am_root->analysis.para.smp_freq;

  /* output file check */
  if (!stout) {
    if (access(filename, F_OK) == 0) {
      if (access(filename, W_OK) == 0) {
	fprintf(stderr, "Warning: overwriting file \"%s\"\n", filename);
      } else {
	perror("mkss");
	return(1);
      }
    }
  }

  /* allocate speech store buffer */
  samples = sfreq * slen / 1000;
  speech = (SP16 *)mymalloc(sizeof(SP16) * samples);

  /* allocate work area to compute spectrum */
  wrk = WMP_work_new(&(jconf->am_root->analysis.para));
  if (wrk == NULL) {
    jlog("ERROR: m_fusion: failed to initialize MFCC computation for SS\n");
    return -1;
  }

  /* initialize input device */
  if (j_adin_init(recog) == FALSE) {
    fprintf(stderr, "Error in initializing adin device\n");
    return -1;
  }

  /* open device */
  if (j_open_stream(recog, NULL) < 0) {
    fprintf(stderr, "Error in opening adin device\n");
  }

  /* record mic input */
  fprintf(stderr, "%dHz recording for %.2f seconds of noise\n", sfreq, (float)slen /(float)1000);
  speechnum = 0;
  adin_go(adin_callback, NULL, recog);

  /* close device */
  adin_end(recog->adin);
  fprintf(stderr, "\n%d samples (%lu bytes, %.1f sec) recorded\n", samples, samples * sizeof(SP16), (float)samples / (float)sfreq);

  /* compute SS */
  fprintf(stderr, "compute SS:\n");
  fprintf(stderr, "  fsize : %4d samples (%.1f msec)\n", jconf->input.framesize, (float)jconf->input.framesize * 1000.0/ (float)sfreq);
  fprintf(stderr, "  fshift: %4d samples (%.1f msec)\n", jconf->input.frameshift, (float)jconf->input.frameshift * 1000.0/ (float)sfreq);

  ss = new_SS_calculate(speech, samples, &sslen, wrk, &(jconf->am_root->analysis.para));

  fprintf(stderr, "  points: %4d\n", sslen);
  fprintf(stderr, "noise spectrum was measured\n");
  
  /* open file for recording */
  fprintf(stderr, "writing average noise spectrum to [%s]...", filename);
  if (stout) {
    fd = 1;
  } else {
    if ((fd = open(filename, O_CREAT | O_RDWR
#ifdef O_BINARY
		   | O_BINARY
#endif
		   , 0644)) == -1) {
      perror("mkss");
      return(1);
    }
  }
  x = sslen;
#ifndef WORDS_BIGENDIAN
  swap_bytes((char *)&x, sizeof(int), 1);
#endif
  if (write(fd, &x, sizeof(int)) < sizeof(int)) {
    perror("mkss");
    return(1);
  }
#ifndef WORDS_BIGENDIAN
  swap_bytes((char *)ss, sizeof(float), sslen);
#endif
  if (write(fd, ss, sslen * sizeof(float)) < sslen * sizeof(float)) {
    perror("mkss");
    return(1);
  }
  if (!stout) {
    if (close(fd) < 0) {
      perror("mkss");
      return(1);
    }
  }
  fprintf(stderr, "done\n");

  WMP_free(wrk);

  return 0;
}

